File: src/main/kotlin/ru/arcam/yggdrasil/service/LinuxController.kt
package ru.arcam.yggdrasil.service
import ru.arcam.yggdrasil.leaf.Leaf
import java.io.BufferedReader
import java.io.InputStreamReader

class LinuxController(leaf: Leaf) : IController(leaf) {
    override fun stop(): String {
        val SERVER_STATUS = arrayOf("sudo", "/usr/bin/systemctl", "stop", leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            return "OK"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "ERROR"
    }
    
    override fun start(): String {
        val SERVER_STATUS = arrayOf("sudo", "/usr/bin/systemctl", "start", leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            return "OK"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "ERROR"
    }

    override fun restart(): String {
        val SERVER_STATUS = arrayOf("sudo", "/usr/bin/systemctl", "restart", leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            return "OK"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "ERROR"
    }
    
    override fun status(): String {
        val SERVER_STATUS = arrayOf("sudo", "/usr/bin/systemctl", "is-active", leaf.name)
        try {
            val process = ProcessBuilder(*SERVER_STATUS).redirectErrorStream(true).start()
            process.waitFor()
            val input = process.inputStream
            val br = BufferedReader(InputStreamReader(input))
            return if (br.readLine() == "active")  "RUNNING" else "STOPPED"
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "UNAVAIVABLE"
    }

    override fun logs(args: List<String>?): String {
        var num = 10
        if (args != null && args.size > 0) {
            val tmp = args[0].toIntOrNull()
            if (tmp != null && tmp > 0)
                num = tmp
        }
        val SERVER_LOGS = arrayOf("sudo", "/usr/bin/journalctl", "--unit=${leaf.name}", "-n", num.toString(), "--no-pager")
        try {
            val process = ProcessBuilder(*SERVER_LOGS).redirectErrorStream(true).start()
            process.waitFor()
            val input = process.inputStream
            val br = BufferedReader(InputStreamReader(input))
            val result = br.readLines().joinToString { it -> "\n" + it }
            return result
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "UNAVAIVABLE"
    }

    override fun callMethod(method: String, args: List<String>): String {
        return "OK"
    }
}

File: service/linux_controller.go
package service

import (
	"os/exec"
	"strings"
	"yggdrasil-branch/models"
)

// LinuxController управляет Linux сервисами через systemctl
type LinuxController struct {
	leaf *models.Leaf
}

// NewLinuxController создает новый LinuxController
func NewLinuxController(leaf *models.Leaf) *LinuxController {
	return &LinuxController{leaf: leaf}
}

// GetLeaf возвращает лист
func (lc *LinuxController) GetLeaf() *models.Leaf {
	return lc.leaf
}

// Stop останавливает сервис
func (lc *LinuxController) Stop() string {
	cmd := exec.Command("sudo", "/usr/bin/systemctl", "stop", lc.leaf.Name)
	if err := cmd.Run(); err != nil {
		return "ERROR"
	}
	return "OK"
}

// Start запускает сервис
func (lc *LinuxController) Start() string {
	cmd := exec.Command("sudo", "/usr/bin/systemctl", "start", lc.leaf.Name)
	if err := cmd.Run(); err != nil {
		return "ERROR"
	}
	return "OK"
}

// Restart перезапускает сервис
func (lc *LinuxController) Restart() string {
	cmd := exec.Command("sudo", "/usr/bin/systemctl", "restart", lc.leaf.Name)
	if err := cmd.Run(); err != nil {
		return "ERROR"
	}
	return "OK"
}

// Status возвращает статус сервиса
func (lc *LinuxController) Status() string {
	cmd := exec.Command("sudo", "/usr/bin/systemctl", "is-active", lc.leaf.Name)
	output, err := cmd.Output()
	if err != nil {
		return "UNAVAIVABLE"
	}
	if strings.TrimSpace(string(output)) == "active" {
		return "RUNNING"
	}
	return "STOPPED"
}

// Logs возвращает логи сервиса
func (lc *LinuxController) Logs(args []string) string {
	num := 10
	if len(args) > 0 {
		if n, err := parseInt(args[0]); err == nil && n > 0 {
			num = n
		}
	}
	
	cmd := exec.Command("sudo", "/usr/bin/journalctl", "--unit="+lc.leaf.Name, "-n", intToString(num), "--no-pager")
	output, err := cmd.Output()
	if err != nil {
		return "UNAVAIVABLE"
	}
	return string(output)
}

// CallMethod вызывает произвольный метод
func (lc *LinuxController) CallMethod(method string, args []string) string {
	return "OK"
}
